File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/methodadapter/tree/TreeMixinPatternMatcher.java
package net.yakclient.mixin.base.internal.methodadapter.tree;

import net.yakclient.mixin.base.internal.instruction.Instruction;
import net.yakclient.mixin.base.internal.instruction.tree.InitialTreeInsnExecutor;
import net.yakclient.mixin.base.internal.instruction.tree.PlacedTreeInsnExecutor;
import net.yakclient.mixin.base.internal.methodadapter.MixinPatternMatcher;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.InsnList;

public abstract class TreeMixinPatternMatcher implements MixinPatternMatcher {
    private final Instruction mixinInsn;

    public TreeMixinPatternMatcher(Instruction instructions) {
        this.mixinInsn = instructions;
    }

    public void insertInsn(AbstractInsnNode index, InsnList list) {
        new PlacedTreeInsnExecutor(list, index, this.mixinInsn).execute();
    }

    public void insertInsn(InsnList list) {
        new InitialTreeInsnExecutor(list, this.mixinInsn).execute();
    }

    public abstract void transform(InsnList insn);
}

File: base/src/main/kotlin/net/yakclient/mixin/base/internal/methodadapter/tree/TreeMixinPatternMatcher.kt
package net.yakclient.mixin.base.internal.methodadapter.tree

import net.yakclient.mixin.base.internal.instruction.Instruction
import net.yakclient.mixin.base.internal.instruction.tree.InitialTreeInsnExecutor
import net.yakclient.mixin.base.internal.instruction.tree.PlacedTreeInsnExecutor
import net.yakclient.mixin.base.internal.methodadapter.MixinPatternMatcher
import org.objectweb.asm.tree.AbstractInsnNode
import org.objectweb.asm.tree.InsnList

abstract class TreeMixinPatternMatcher(private val mixinInsn: Instruction?) : MixinPatternMatcher {
    fun insertInsn(index: AbstractInsnNode?, list: InsnList) {
        if (index != null) PlacedTreeInsnExecutor(list, index, mixinInsn!!).execute()
    }

    fun insertInsn(list: InsnList?) {
        InitialTreeInsnExecutor(list!!, mixinInsn!!).execute()
    }

    abstract fun transform(insn: InsnList)
}